dashboard_cache.php

```diff
@@ -7,8 +7,13 @@
 $project_id = (int)$_GET['pid'];
 $report = htmlentities($_GET['report'],ENT_QUOTES);
 $banner = $module->getProjectSetting('banner',$project_id);
+$path = $module->getProjectSetting('path',$project_id);
```

Comment:
This likely warrants a getSafePath() call as well.  It would also likely be prudent to only use one getSafePath() in a shared method referenced in all locations where this path setting is accessed (as opposed to duplicating the getSafePath() call, which would be easy to miss during future changes).

Reply:
The problem with this path is that we don't know which root they will use. To my understanding, getSafePath needs a root specified if outside the module.
This is for an institution and they want to use their own path and this one not necessarily being inside the module (and definitely not in edocs). We talked with them about security issues but it was decided to let them have their own path.
Any recommendations in here?

Reply:
What would some example paths be?  I don't think we want to allow users to specify any root on the system, as that would be effectively supporting file path injection vulnerabilities.  There has to be some way we can limit it.

Reply:
I'll ask Alex. I don't know exactly what they want to use as a path.

Reply:
So it seems they use Amazon S3 buckets. Completely different of just simply a path...
I'm on standby to see what they want to do.
I might just do a new branch for this 
https://github.com/vanderbilt-redcap/dashboard-analysis-platform/pull/19/commits/bbf69467221649064c7c928ab8fbbb16d36b3f41 
while they decide on what to do

Reply:
If the path is actually an S3 url & not a file path, then it could be made safe by reporting an error or ignoring any values users that don't start with `s3://`.

Reply:
but if it's an S3 url would the getSafePath still work to save the file? And I'm guessing the root would be `s3://`.
I like the idea of ignoring anything that is not an S3 this way we won't allow everyone to use this feature.

Reply:
The getSafePath() method only works on actual file URLs.  I think determining whether user sourced parameters are safe to use for building s3 URLs is going to have to be considered on a case by case basis.  I can imaging some cases where that is safe, and some where it is not.

Reply:
As far as I know only this institution will have S3 urls and it will be on their own server.

Reply:
Also my main concern is on how to test it as I don't think they will give us access.

Reply:
We should still ensure these urls are prefixed with `s3://` before calling `file_get_contents()` on them.

Reply:
Oh yes, I'm definitely doing that and not allowing anything else. I'll work on it and update the branch.

Reply:
Thank you for the suggestions!

Reply:
Just made an update. Let me know your thoughts!

Reply:
Looks great to me!

Reply:
I apologize, I did not look at this properly on the first pass.  This logic should be placed in a shared function and reused rather than copy pasting.  Copy pasting and duplicating code in general is very bad practice and creates significant technical debt.  Please stop doing this in general.  I would recommend a module function like this, to remove duplication of the `getProjectSetting()` call as well:
```php
    function getS3Path(){
            $path = $this->getProjectSetting('path',$project_id);

            if (stripos($path, "s3://") === 0) {
                //It matches
            }else{
                $path = null;
            }
            
            return $path;
    }
```

Reply:
You are 100% right. I went ahead and did that and added some other improvements.
It's a 2x1  deal ;)
Thank you for taking your time to check on this and for your useful suggestions.